Enable sparse workload for ctl read

xinlian12 opened the pull request:
Enable --SparsityWaitTime support for ctlworkload read operation.

moderakh reviewed (commented):
Thanks for adding this Annie.

please be aware with sparse workload the warm up time is high, that was the reason I introduced `skipWarmUpOperations`. So it is worth utilizing the sparse workload with `skipWarmUpOperations` option.

moderakh commented on sdk/cosmos/azure-cosmos-benchmark/src/main/java/com/azure/cosmos/benchmark/ctl/AsyncCtlWorkload.java:
```diff
@@ -172,10 +173,14 @@ private void performWorkload(BaseSubscriber<Object> documentSubscriber, Operatio
-                .flux();
+
+            Mono sparsitySleepMono = sparsityMono();
+            obs = sparsitySleepMono.flux().doOnComplete(
```
any reason we are not doing flatmap similar to the original AsyncReadBenchmark?
https://github.com/Azure/azure-sdk-for-java/blob/master/sdk/cosmos/azure-cosmos-benchmark/src/main/java/com/azure/cosmos/benchmark/AsyncReadBenchmark.java#L84

xinlian12 replied:
The reason being that the sparsitySleepMono in ctlworkload returns Mono.delay or Mono.empty
sparsitySleepMono in AsyncReadBenchmark returns Mono.delay or null. And only flatmap when Mono.delay. 

So i changed to doOnComplete instead. But let me know this is a bad choice I will change~

moderakh replied:
either way not a big deal. however usually flatMap is used for representing a chain of operations. in terms of correctness what you are doing is correct as well.

You can use flatmap to chain on completion.

xinlian12 replied:
thanks Mo, I changed back to the way you had before except that instead of null return Duration.ofSeconds(0).

moderakh reviewed (approved):
looks good to me. I added two minor comments.

/CC @simplynaveen20 as this is CTL.

xinlian12 commented:
> Thanks for adding this Annie.
> 
> please be aware with sparse workload the warm up time is high, that was the reason I introduced `skipWarmUpOperations`. So it is worth utilizing the spark workload with `skipWarmUpOperations` option.

thanks Mo~ 
Is the 'skipWarmUpOperations` main purpose is that we can calculate the latency more accurate?  I did not enable this option because the following two thinkings:
1. Sometime back i changed the underlying histogram we use for latencyMeter, so after every report snapshot, the data will be reset. Currently ctl work load is using default 10s.
2. Since usually each ctl run will take about several hours, so I think it should be fine for us to just ignore the first 30s or so when we compare the perf data.

simplynaveen20 commented on sdk/cosmos/azure-cosmos-benchmark/src/main/java/com/azure/cosmos/benchmark/ctl/AsyncCtlWorkload.java:
```diff
@@ -370,4 +375,13 @@ private void initializeReporter(Configuration configuration) {
+
+    protected Mono sparsityMono() {
+        Duration duration = configuration.getSparsityWaitTime();
+        if (duration != null && !duration.isZero()) {
```
Why no skip warmup operation like in regular bench mark ? Also can we make is a static method as one place and user across places ?

xinlian12 replied:
Yea agree that is a good point.

So after comparing AsyncBenchmark and AsyncCtlWorkload, I see there are a lot common methods, like creating cosmosClient, create database, initialize reporter etc. And since they all part of the benchmark, so I think it makes sense to have a base class to concentralize all those common methods. 

So I created a BenchmarkBase class instead. And moved some methods to the base class including the sparsityMono and initializeMetersIfSkippedEnoughOperations. 

@moderakh @simplynaveen20 @kushagraThapar  Can u take a look again since there are some refactor since last time, thanks~ (and let me know if I have gone toooo far...)

simplynaveen20 commented on sdk/cosmos/azure-cosmos-benchmark/src/main/java/com/azure/cosmos/benchmark/ctl/AsyncCtlWorkload.java:
```diff
@@ -172,10 +173,14 @@ private void performWorkload(BaseSubscriber<Object> documentSubscriber, Operatio
-                PojoizedJson.class)
-                .flux();
+
+            Mono sparsitySleepMono = sparsityMono();
```
Why only in read operation , don't we need in write and query , as they all part of on CTL workload

xinlian12 replied:
Yea, make sense Naveen, changed to all operations in CTL worklaod (write,query, read)

xinlian12 commented:
> Thanks for adding this Annie.
> 
> please be aware with sparse workload the warm up time is high, that was the reason I introduced `skipWarmUpOperations`. So it is worth utilizing the spark workload with `skipWarmUpOperations` option.

Thanks Mo, good catch, updated in the new iterator

FabianMeiswinkel reviewed (approved):
LKTM - thanks for the refactoring!

moderakh commented on sdk/cosmos/azure-cosmos-benchmark/src/main/java/com/azure/cosmos/benchmark/AsyncReadBenchmark.java:
```diff
@@ -78,34 +74,26 @@ protected void performWorkload(BaseSubscriber<PojoizedJson> baseSubscriber, long
     private void readThroughput(Mono<PojoizedJson> readItem, BaseSubscriber<PojoizedJson> baseSubscriber, long i) {
         Mono sparsitySleepMono = sparsityMono(i);
-        Mono<PojoizedJson> result = readItem;
-        if (sparsitySleepMono != null) {
```
the reason initially I decided to go with null value for `sparsitySleepMono` was because it would help to bypass the flatmap. I suspect you will have a ~1-2% perf hit as a result of this change on the normal path.

could you do a perf run using this benchmark tool both on the master and on your PR and measure the perf difference?

JonathanGiles commented:
I'm curious if this PR is able to be merged?

xinlian12 commented:
@JonathanGiles will need to retest the change.
